examples/simple_trainer_mip_splatting.py

```diff
@@ -0,0 +1,1094 @@
+                        is_coalesced=len(Ks) == 1,
+                    )
+
+            # optimize
+            for optimizer in self.optimizers:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+            for optimizer in self.pose_optimizers:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+            for optimizer in self.app_optimizers:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+            for scheduler in schedulers:
+                scheduler.step()
+
+            # save checkpoint
+            if step in [i - 1 for i in cfg.save_steps] or step == max_steps - 1:
+                mem = torch.cuda.max_memory_allocated() / 1024**3
+                stats = {
+                    "mem": mem,
+                    "ellipse_time": time.time() - global_tic,
+                    "num_GS": len(self.splats["means3d"]),
+                }
+                print("Step: ", step, stats)
+                with open(f"{self.stats_dir}/train_step{step:04d}.json", "w") as f:
+                    json.dump(stats, f)
+                torch.save(
+                    {
+                        "step": step,
+                        "splats": self.splats.state_dict(),
+                    },
+                    f"{self.ckpt_dir}/ckpt_{step}.pt",
+                )
+
+            # eval the full set
+            if step in [i - 1 for i in cfg.eval_steps] or step == max_steps - 1:
+                self.eval(step)
+                self.render_traj(step)
+
+            if not cfg.disable_viewer:
+                self.viewer.lock.release()
+                num_train_steps_per_sec = 1.0 / (time.time() - tic)
+                num_train_rays_per_sec = (
+                    num_train_rays_per_step * num_train_steps_per_sec
+                )
+                # Update the viewer state.
+                self.viewer.state.num_train_rays_per_sec = num_train_rays_per_sec
+                # Update the scene.
+                self.viewer.update(step, num_train_rays_per_step)
+
+    @torch.no_grad()
+    def compute_3D_smoothing_filter(self):
+        cfg = self.cfg
+        device = self.device
+        xyz = self.splats["means3d"]
+        print("xyz", xyz.shape, xyz.device)
+
+        worldtocams = (
+            torch.from_numpy(self.trainset.parser.worldtocams).float().to(device)
+        )
+        # TODO, currently use K, H, W of the first image
+        data = self.trainset[0]
+        K = data["K"].to(device)  # [3, 3]
+        height, width = data["image"].shape[:2]
+        Ks = torch.stack([K] * len(worldtocams), dim=0)  # [C, 3, 3]
+        filter_3D = compute_3D_smoothing_filter(
+            xyz, worldtocams, Ks, width, height, cfg.near_plane
+        )
+        filter_3D = filter_3D * (0.2**0.5)
+        self.splats["filters"] = torch.nn.Parameter(filter_3D)
+
+    @torch.no_grad()
+    def compute_3D_smoothing_filter_torch(self):
```

Comment:
Would you mind move this into `gsplat/cuda/_torch_impl.py` and put up a test case under `test/test_basic.py` to protect the CUDA implementation `compute_3D_smoothing_filter()`?